src/modules/rlm_sql/drivers/rlm_sql_db2/rlm_sql_db2.c

```diff
@@ -106,50 +106,38 @@ static sql_rcode_t sql_socket_init(rlm_sql_handle_t *handle, rlm_sql_config_t co
 	return RLM_SQL_OK;
 }
 
-static sql_rcode_t sql_query(rlm_sql_handle_t *handle, UNUSED rlm_sql_config_t const *config, char const *query)
+static unlang_action_t sql_query(rlm_rcode_t *p_result, UNUSED int *priority, request_t *request, void *uctx)
 {
+	fr_sql_query_t		*query_ctx = talloc_get_type_abort(uctx, fr_sql_query_t);
 	SQLRETURN row;
 	rlm_sql_db2_conn_t *conn;
 
-	conn = handle->conn;
+	conn = query_ctx->handle->conn;
 
 	/* allocate handle for statement */
 	SQLAllocHandle(SQL_HANDLE_STMT, conn->dbc_handle, &(conn->stmt));
 
 	/* execute query */
 	{
 		SQLCHAR *db2_query;
-		memcpy(&db2_query, &query, sizeof(query));
+		memcpy(&db2_query, &query_ctx->query_str, sizeof(query_ctx->query_str));
 
 		row = SQLExecDirect(conn->stmt, db2_query, SQL_NTS);
 		if(row != SQL_SUCCESS) {
 			/* XXX Check if row means we should return RLM_SQL_RECONNECT */
 			ERROR("Could not execute statement \"%s\"", query);
-			return RLM_SQL_ERROR;
+			query_ctx->rcode = RLM_SQL_ERROR;
+			RETURN_MODULE_FAIL
 		}
 	}
 
-	return RLM_SQL_OK;
-}
-
-static sql_rcode_t sql_select_query(rlm_sql_handle_t *handle, rlm_sql_config_t const *config, char const *query)
-{
-	return sql_query(handle, config, query);
-}
-
-static int sql_num_fields(rlm_sql_handle_t *handle, UNUSED rlm_sql_config_t const *config)
-{
-	SQLSMALLINT c;
-	rlm_sql_db2_conn_t *conn;
-
-	conn = handle->conn;
-	SQLNumResultCols(conn->stmt, &c);
-	return c;
+	query_ctx->rcode = RLM_SQL_OK;
+	RETURN_MODULE_OK;
 }
 
-static sql_rcode_t sql_fields(char const **out[], rlm_sql_handle_t *handle, UNUSED rlm_sql_config_t const *config)
+static sql_rcode_t sql_fields(char const **out[], fr_sql_query_t *query_ctx, UNUSED rlm_sql_config_t const *config)
```

Comment:
Should really be `rlm_sql_rcode_t` and `rlm_sql_query_t`, or `sql_rcode_t` and `sql_query_t` the `fr_` prefix is meant to be reserved for the utility and protocol libraries to allow reuse outside of freeradius.